src/Core/Aetherling/Passes/ThroughputModifications.hs

```diff
@@ -1,339 +1,508 @@
+attemptSpeedUp requestedMult op@(Not _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(And _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Or  _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(XOr _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Eq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Neq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Lt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Leq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Gt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Geq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(LUT _) = (MapOp requestedMult op, requestedMult)
+
+-- Must map over these because making a single memory wider is a more
+-- expensive operation than banking it (making multiple copies)
+-- NOTE: unclear what expensive means here, just my understanding from
+-- talking with Spatial team, Ross, and other hardware people. I should
+-- figure it out.
+-- maping over a MemRead/MemWrite is banking it, and wrapping
+-- the type with an array is making a single memory read/write more
+-- per clock.
+attemptSpeedUp requestedMult op@(MemRead _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(MemWrite _) = (MapOp requestedMult op, requestedMult)
+
+-- These are leaf, non-modifiable rate unlike SequenceArrayRepack because,
+-- for these operations, the user has not specified the type separately from the
+-- array length. Therefore, speedup may modify the meaning of the program
+-- by changing the types of these ops.
+attemptSpeedUp requestedMult op@(ArrayReshape _ _) =
+  (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(DuplicateOutputs _ _) =
+  (MapOp requestedMult op, requestedMult)
+
+attemptSpeedUp requestedMult op@(Constant_Int _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Constant_Bit _) = (MapOp requestedMult op, requestedMult)
+
+-- LEAF, MODIFIABLE RATE
+-- If possible, speed up by changing the rate. Otherwise, just try to return
+-- an op with the same or greater throughput than the original.
+-- However, behavior is undefined if requesting an invalid throughput multiplier
+-- that doesn't match the divisibility requirements such as those for linebuffer.
+-- Not going to speed these up by wrapping them in a map because that would
+-- cause inappropriate behavior in the linebuffer. The linebuffer has complex
+-- internal state and mapping over it will cause it to behave unintuitively.
+-- For example, an LineBuffer [2] [3] [300] T_Int will reading in 2 int per clock
+-- and emit two overlapping windows of consecutive ints.
+-- A Map 2 (LineBuffer [1] [3] [300] T_Int) will emit two windows, but the
```

Comment:
We should be careful here.  There is an assumption in the comment about the behavior of prior pipeline stages.  It's that the stream of tokens arriving at the line buffer was not reshuffled accordingly.  (This is a reasonable assumption, but it's not necessarily true that the stream arriving at the linebuffer was the same stream as before the speedup, just vectorized.)

Reply:
My design for speedup and slowdown actually does assume that the streams are not modified excepting number of bits processed per clock. I've made this assumption explicit: https://github.com/David-Durst/aetherlingHaskellIR/pull/14/files#diff-a4e9be19161290150b0e024cdc796b07R5